subprojects/gradle/build-trace/src/gradleTest/kotlin/com/avito/android/build_trace/CriticalPathTest.kt

```diff
@@ -0,0 +1,253 @@
+package com.avito.android.build_trace
+
+import com.avito.android.build_trace.internal.critical_path.CriticalPathSerialization
+import com.avito.test.gradle.TestResult
+import com.avito.test.gradle.gradlew
+import com.google.common.truth.Truth.assertThat
+import org.gradle.testkit.runner.TaskOutcome
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Disabled
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+
+class CriticalPathTest {
+
+    private lateinit var projectDir: File
+
+    @BeforeEach
+    fun setup(@TempDir tempDir: File) {
+        this.projectDir = tempDir
+    }
+
+    @Test
+    fun `has a predecessor task - input-output`() {
+        setupTasks(
+            """
+            abstract class ProducerTask @Inject constructor(objects: ObjectFactory) : DefaultTask() {
+            
+                @OutputFile
+                val outputFile = objects.fileProperty()
+            
+                @TaskAction
+                fun createFile() {
+                    outputFile.get().asFile.writeText("Output of a producer")
+                }
+            }
+            
+            abstract class ConsumerTask @Inject constructor(objects: ObjectFactory) : DefaultTask() {
+            
+                @InputFile
+                var input = objects.fileProperty()
+            
+                @OutputFile
+                val outputFile = objects.fileProperty()
+            
+                @TaskAction
+                fun createFile() {
+                    outputFile.get().asFile.writeText("Output of a consumer + " + input.get().asFile.readText())
+                }
+            }
+            
+            val producer = tasks.register("producerTask", ProducerTask::class.java) {
+                outputFile.set(file("build/producer.txt"))
+            }
+            
+            tasks.register("consumerTask", ConsumerTask::class.java) {
+                input.set(producer.get().outputFile)
+                outputFile.set(file("build/consumer.txt"))
+            }
+            """.trimIndent()
+        )
+
+        val tasks = calculatePath(":consumerTask")
+
+        assertThat(tasks).containsExactly(":producerTask", ":consumerTask")
+    }
+
+    @Test
+    fun `has a predecessor task - dependsOn`() {
```

Comment:
As I see, we use mostly this pattern `state under test - expected behaviour - [when]`.

Samples:

```kotlin
@Test
fun `success runs - returns number of passed test runs`() {}

@Test
fun `read instrumentation output - completes stream - with failed test`() {}

@Test
fun `create summary - marks test as matched - all tests passed`() {}
```
 
I tend to think this way:
- "critical path" - state under test. I've omitted this part because of repeating.
-  "has a predecessor task" - expected behavior. This is ok.
- "dependsOn" - when part. This is what confuses me. I don't know to interpret this - as a condition or as a subject of testing.

Agree that "dependsOn" is what we really test here, in the wider context of a critical path.